### PR TITLE
Add support for nested locks in MySQL

### LINKED ISSUE #29: MySQL supports nested locks in >= 5.7.5
https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_get-lock

Dunno if you want to complicate the code by adding support for nested locks in that case.. but might be nice.

### PR DESCRIPTION
**Pull request by atyndall**
This PR introduces support for nested locks in MySQL, which were introduced in MySQL 5.7.5.
More info in PR code comments.

Fixes #29.

### CODE REVIEW

**Comment by atyndall** [comment]
Currently working on getting tests passing

**Comment by atyndall** [comment]
@mceachen should be ready for review now

I'm having one problem with one specific test failure; https://travis-ci.org/ClosureTree/with_advisory_lock/jobs/413841405
Have you seen this problem before?

**Comment by mceachen** [comment]
> Have you seen this problem before?
No, but it's on ruby 2.2, so let's kill that from the build matrix and call it a day. Great PR, thanks for your work!

**Comment by atyndall** [comment]
Rebased over new master

**Comment by mceachen** [comment]
@seuros if you'd like to review, that'd be great. If you don't have time, I'll merge this on Friday.

**Comment by mceachen** [comment]
@atyndall thanks again for the PR! Great work.

**Comment by atyndall** [comment]
Hmm, now a postgres test is failing; is this a known issue with master @mceachen, or should I investigate it as a possible weird regression caused by my code?

**Comment by mceachen** [comment]
I'll take a look tomorrow. Unfortunately, Travis can be flaky, and these
big build matrixes exacerbate the flaketasticness.

File: lib/with_advisory_lock/concern.rb [hunk 1]
```diff
@@ -33,7 +33,11 @@ def impl_class
         if adapter.postgresql?
           WithAdvisoryLock::PostgreSQL
         elsif adapter.mysql?
-          WithAdvisoryLock::MySQL
+          if adapter.mysql_nested_lock_support?
```
Comment (seuros) [inline lib/with_advisory_lock/concern.rb]: We need a opt-out configuration to not use nested locks even if the adapter support it.

**Comment by atyndall** [comment]
@seuros I've made the change you've requested. Open to further suggestions on naming / how the options should operate.

**Comment by seuros** [comment]
@atyndall Thank you for your contribution

**Pull request merged**